src/logic/__tests__/startAccountMigration.test.js

```diff
@@ -0,0 +1,107 @@
+import {Observable} from 'rxjs';
+
+import startAccountMigration from '../startAccountMigration';
+import {
+  accountMigrationComplete,
+  accountMigrationError,
+  accountMigrationUndoPeriodExpired,
+} from '../../actions/user';
+import {migrateAccount} from '../../clients/firebase';
+import {getCurrentAccountMigration} from '../../selectors';
+import {bugsnagClient} from '../../util/bugsnag';
+
+import {
+  credentialFactory,
+  firebaseErrorFactory,
+  userFactory,
+} from '@factories/clients/firebase';
+
+import {
+  githubRepositoryFactory,
+} from '@factories/clients/github';
+
+
+jest.mock('../../actions/user.js');
+jest.mock('../../clients/firebase.js');
+jest.mock('../../selectors');
+jest.mock('../../util/bugsnag.js');
+
+const getState = jest.fn();
+const dispatch = jest.fn();
+const done = jest.fn();
+
+describe('startAccountMigration', () => {
+  test('not dismissed during undo period, successful migration', async() => {
+    const mockCredential = credentialFactory.build();
+    getCurrentAccountMigration.mockResolvedValue({
+      firebaseCredential: mockCredential,
+    });
+
+    const mockUser = userFactory.build();
+    const mockProjects = [
+      githubRepositoryFactory.build(),
+      githubRepositoryFactory.build(),
+    ];
+    migrateAccount.mockResolvedValue({
+      user: mockUser,
+      migratedProjects: mockProjects,
+    });
+
+    const emptyDispatch = new Observable((subscriber) => {
+      subscriber.complete();
```

Comment:
I don’t think the `action$` observable would ever be expected to complete in the real application—that would essentially imply that there will be no more actions dispatched to Redux, which isn’t a situation that would ever obtain. So I think this test is actually simulating an impossible situation, specifically that `cancelPromise` in the logic resolves to a `false` value.

I believe the situation we want to be testing here (the user does not cancel the migration) is that the `continuePromise` resolves to `true`, which would mean having this observable be _truly_ empty (does not dispatch or complete), and then use a [jest timer mock](https://jestjs.io/docs/en/timer-mocks) to simulate the timeout completing without making the test actually sit there for five seconds.

Reply:
May need help on this one, currently seems that Promises and timer mocks don't play well.
Used solution here: https://stackoverflow.com/questions/51126786/jest-fake-timers-with-promises, but can't seem to get it to use await sync properly.

Reply:
Can you push the non-working test so I can play around with it?

Reply:
One thing that I imagine might not be intuitive is that you will need to hold a reference to the promise; then advance the timers; then await the promise, something like:

```js
const migrationDone = startAccountMigration.process(
  {action$: emptyDispatch, getState},
  dispatch,
   done,
);
jest.runAllTimers();
await migrationDone;
```

Not sure if that’s helpful but figured I’d throw it out there.

Reply:
Ah yeah that actually works out.